src/AzureIoTHub.Portal.Server/Services/LoRaWanDeviceService.cs

```diff
@@ -202,20 +220,18 @@ public override async Task ProcessTelemetryEvent(EventData eventMessage)
             return;
         }
 
-        private async Task KeepOnlyLatestHundredTelemetry(LorawanDevice loRaWanDevice)
+        private static void KeepOnlyLatestTelemetry(LorawanDevice loRaWanDevice, int numberOfMessages = 100)
         {
-            if (loRaWanDevice.Telemetry.Count <= 100) return;
+            if (loRaWanDevice.Telemetry.Count <= numberOfMessages) return;
 
             loRaWanDevice.Telemetry
                 .OrderByDescending(telemetry => telemetry.EnqueuedTime)
-                .Skip(100)
+                .Skip(numberOfMessages)
                 .ToList()
                 .ForEach(telemetry =>
                 {
-                    this.deviceTelemetryRepository.Delete(telemetry.Id);
+                    _ = loRaWanDevice.Telemetry.Remove(telemetry);
```

Comment:
Telemetry entity is removed from lorawan device but not from database

Reply:
Maybe that's why you used SaveAsync twice during this action : 
- Removing from the collection
- removing from the table

You might have another option by setting non-nullable deviceId on the telemetry entity and specify that the context should delete the entity when removed from the collection ?

Reply:
We can move the method KeepOnlyLatestTelemetry to LoRaDeviceTelemetryRepository and use the DbContext to update the state of each telemetry entity. Example (not tested 😄):

```csharp
public void KeepOnlyLatestTelemetry(LorawanDevice loRaWanDevice, , int numberOfMessages = 100)
{
    loRaWanDevice.Telemetry
        .OrderByDescending(telemetry => telemetry.EnqueuedTime)
        .Skip(numberOfMessages)
        .ToList()
        .ForEach(telemetry =>
        {
            this.context.Entry(telemetry).State = Microsoft.EntityFrameworkCore.EntityState.Deleted;
        });
}
```

What do you think?

Reply:
Context is not available in the services. 
I re-added the repository call to delete method